spark/src/main/java/org/apache/iceberg/spark/source/IcebergTable.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.source;
+
+import com.google.common.collect.ImmutableSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.iceberg.PartitionField;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.spark.SparkSchemaUtil;
+import org.apache.iceberg.spark.SparkUtils;
+import org.apache.iceberg.transforms.UnknownTransform;
+import org.apache.iceberg.types.CheckCompatibility;
+import org.apache.spark.sql.connector.catalog.SupportsRead;
+import org.apache.spark.sql.connector.catalog.SupportsWrite;
+import org.apache.spark.sql.connector.catalog.Table;
+import org.apache.spark.sql.connector.catalog.TableCapability;
+import org.apache.spark.sql.connector.expressions.Transform;
+import org.apache.spark.sql.connector.read.ScanBuilder;
+import org.apache.spark.sql.connector.write.BatchWrite;
+import org.apache.spark.sql.connector.write.SupportsDynamicOverwrite;
+import org.apache.spark.sql.connector.write.SupportsTruncate;
+import org.apache.spark.sql.connector.write.WriteBuilder;
+import org.apache.spark.sql.connector.write.streaming.StreamingWrite;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.util.CaseInsensitiveStringMap;
+
+final class IcebergTable implements Table, SupportsRead, SupportsWrite {
+
+  private static final Set<TableCapability> CAPABILITIES = ImmutableSet.of(
+      TableCapability.BATCH_READ,
+      TableCapability.BATCH_WRITE,
+      TableCapability.MICRO_BATCH_READ,
+      TableCapability.STREAMING_WRITE,
+      TableCapability.TRUNCATE,
```

Comment:
This should support `TableCapability.OVERWRITE_BY_FILTER` instead of truncate.

Reply:
IIUC, with the current code, we cannot support `OVERWRITE_BY_FILTER`. Overwrite by filter requires to delete the rows which match the filter, but the current Iceberg API only supports file based deletion by filter, not row based.

Reply:
The API allows us to reject filters that require row-level deletes. Spark's overwrite mostly uses partition filters anyway, so it works.

Reply:
I see, thanks for explanation. let me check it.

Reply:
I should clarify: Spark uses partition filters when running `INSERT OVERWRITE` on a partitioned table with `static` mode. In dynamic mode, it should call the dynamic overwrite.

Static mode drops partitions identified by the static data in the `PARTITION` clause. For example,

```sql
INSERT OVERWRITE t PARTITION (log_level='TRACE', log_date) SELECT ...
```

That will pass a delete filter `log_level = 'TRACE'`. Because these come from the `PARTITION` clause, they are safe to run.